File: tsnode/flex-fields.ts
import * as api from './api/apis';
import * as model from './model/models';
import {Configuration} from './runtime';

export class FlexFields
{
    private _configuration: Configuration;

    private _request: model.InitiateInstallmentPlanRequest;
    private _api?: api.InstallmentPlanApi;

    constructor(configuration: Configuration){
        this._configuration = configuration;

        this._request = {
            planData: {}
        };
    }

    public static async authenticate(env: Configuration, username: string, password: string, apiKey?: string) : Promise<FlexFields> {
        if (apiKey){
            env.addApiKey(apiKey!);
        }

        let ff = new FlexFields(env);
        let sessionId = await ff.getSessionId(username, password);
        ff._api = new api.InstallmentPlanApi(env, sessionId);
        return ff;
    }

    public addCulture(culture: string){
        this._api!.setCulture(culture);
        return this;
    }

    public addInstallments(installmentOptions: Array<number>, defaultNumInstallments?: number) : FlexFields
    {
        this._request.planData!.numberOfInstallments = defaultNumInstallments;

        if (this._request.paymentWizardData == null)
        {
            this._request.paymentWizardData = <model.PaymentWizardData>{};
        }

        this._request.paymentWizardData.requestedNumberOfInstallments = installmentOptions.join(',');
        return this;
    }

    public addBillingInformation(billingAddress: model.AddressData, consumerData: model.ConsumerData) : FlexFields
    {
        this._request.billingAddress = billingAddress;
        this._request.consumerData = consumerData;
        return this;
    }

    public add3DSecure(redirectUrls: model.RedirectUrls) : FlexFields
    {
        let attempt3d = <any>true;
        this._request.planData!.attempt3DSecure = attempt3d;
        this._request.redirectUrls = redirectUrls;
        return this;
    }

    public addCaptureSettings(
        autoCapture?: boolean, 
        firstInstallmentAmount?: number,
        currencyCode?: string,
        firstChargeDate?: Date) : FlexFields
    {
        this._request.planData!.autoCapture = autoCapture;
        
        if (firstInstallmentAmount != null){
            this._request.planData!.firstInstallmentAmount = <model.MoneyWithCurrencyCode>{ value: firstInstallmentAmount, currencyCode: currencyCode };
        }
        
        this._request.planData!.firstChargeDate = firstChargeDate;
        return this;
    }

    public async getPublicToken(amount: number, currencyCode: string) : Promise<string> {
        this._request.planData!.amount = <model.MoneyWithCurrencyCode>{ value: amount, currencyCode: currencyCode };
        var initPlanResponse = await this._api!.installmentPlanInitiate(this._request);

        return initPlanResponse.body.publicToken!;
    }

    public async verifyPayment(planNumber: string, orderAmount: number) : Promise<boolean>
    {
        var paymentVerificationResponse = await this._api!.installmentPlanVerifyPayment(<model.VerifyPaymentRequest>{ 
            installmentPlanNumber: planNumber 
        });
        
        if (paymentVerificationResponse.body.isPaid && paymentVerificationResponse.body.originalAmountPaid == orderAmount)
        {
            return true;
        }

        return false;
    }


    private async getSessionId(username: string, pass: string) : Promise<string> {
        var loginApi = new api.LoginApi(this._configuration!);
        var request = <model.LoginRequest>{ userName: username, password: pass };

        var loginResult = await loginApi.loginPost(request);
        return loginResult.body.sessionId!;
    }
}

File: python-async/splitit_async/flex_fields.py
import splitit_async


class FlexFields(object):
    def __init__(self, api_client):
        self.api_client = api_client
        self._active_init_req = splitit_async.InitiateInstallmentPlanRequest(
            plan_data=splitit_async.PlanData()
        )

    async def __aenter__(self):
        await self.api_client.__aenter__()
        return self

    async def __aexit__(self, exc_type, exc, tb):
        await self.api_client.__aexit__(exc_type, exc, tb)

    def add_installments(self, installment_options, default_num_installments):
        self._active_init_req.plan_data.number_of_installments = default_num_installments
        if self._active_init_req.payment_wizard_data is None:
            self._active_init_req.payment_wizard_data = splitit_async.PaymentWizardData()

        self._active_init_req.payment_wizard_data.requested_number_of_installments = installment_options

    def add_billing_information(self, billing_address, consumer_data):
        self._active_init_req.billing_address = billing_address
        self._active_init_req.consumer_data = consumer_data

    async def get_public_token(self, amount, currency_code):
        self._active_init_req.plan_data.amount = splitit_async.MoneyWithCurrencyCode(
            amount, currency_code
        )

        resp = await self.api_client.InstallmentPlanApi.installment_plan_initiate(
            self._active_init_req
        )

        return resp.public_token

    async def verify_payment(self, plan_number, order_amount):
        resp = await self.api_client.InstallmentPlanApi.installment_plan_verify_payment(
            splitit_async.VerifyPaymentRequest(
                plan_number
            )
        )
        return resp.is_paid and resp.original_amount_paid == order_amount

    def add_3D_secure(self, redirect_urls):
        self._active_init_req.plan_data.attempt3_d_secure = True
        self._active_init_req.redirect_urls = redirect_urls

    def add_capture_settings(self, auto_capture=False, first_installment_amount=0, first_charge_date=None):
        self._active_init_req.plan_data.auto_capture = auto_capture
        if first_installment_amount > 0:
            self._active_init_req.plan_data.first_installment_amount = first_installment_amount
        if first_charge_date:
            self._active_init_req.plan_data.first_charge_date = first_charge_date
